packages/micropip/src/micropip/micropip.py

```diff
@@ -185,19 +190,25 @@ async def install(
         if len(pyodide_packages):
             # Note: branch never happens in out-of-browser testing because in
             # that case BUILTIN_PACKAGES is empty.
-            self.installed_packages.update(pyodide_packages)
+            self.installed_packages.update({pkg.name: pkg for pkg in pyodide_packages})
             wheel_promises.append(
                 asyncio.ensure_future(
                     pyodide_js.loadPackage(
-                        to_js([name for [name, _] in pyodide_packages])
+                        to_js([name for [name, _, _] in pyodide_packages])
                     )
                 )
             )
 
         # Now install PyPI packages
         for name, wheel, ver in transaction["wheels"]:
             wheel_promises.append(_install_wheel(name, wheel))
-            self.installed_packages[name] = ver
+
+            # detect whether the wheel metadata is from PyPI or from custom location
+            # wheel metadata from PyPI has SHA256 checksum digest.
+            wheel_source = "pypi" if wheel["digests"] is not None else wheel["url"]
```

Comment:
This (checking digest) is a hacky way to detect whether the wheel is from PyPI or from a custom URL.
Another way can be adding a key that explicitly indicates the wheel is from a custom URL.